model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RealmCacheSession.java

```diff
@@ -546,6 +546,42 @@ static String getRoleByNameCacheKey(String container, String name) {
         return getClientDelegate().getClients(realm);
     }
 
+    @Override
+    public List<ClientModel> getAlwaysDisplayInConsoleClients(RealmModel realm) {
+        String cacheKey = getRealmClientsQueryCacheKey(realm.getId());
+        boolean queryDB = invalidations.contains(cacheKey) || listInvalidations.contains(realm.getId());
+        if (queryDB) {
+            return getClientDelegate().getAlwaysDisplayInConsoleClients(realm);
+        }
+
+        ClientListQuery query = cache.get(cacheKey, ClientListQuery.class);
+        if (query != null) {
+            logger.tracev("getClients cache hit: {0}", realm.getName());
+        }
+
+        if (query == null) {
+            Long loaded = cache.getCurrentRevision(cacheKey);
+            List<ClientModel> model = getClientDelegate().getAlwaysDisplayInConsoleClients(realm);
+            if (model == null) return null;
+            Set<String> ids = new HashSet<>();
+            for (ClientModel client : model) ids.add(client.getId());
+            query = new ClientListQuery(loaded, cacheKey, realm, ids);
+            logger.tracev("adding realm clients cache miss: realm {0} key {1}", realm.getName(), cacheKey);
+            cache.addRevisioned(query, startupRevision);
+            return model;
+        }
+        List<ClientModel> list = new LinkedList<>();
+        for (String id : query.getClients()) {
+            ClientModel client = session.realms().getClientById(id, realm);
```

Comment:
If I understand this correctly, we don't fully cache the results? Because here we're again accessing `ClientModel`.